docs/sql-migration-guide-upgrade.md

```diff
@@ -37,6 +37,8 @@ displayTitle: Spark SQL Upgrading Guide
 
   - In Spark version 2.4 and earlier, CSV datasource converts a malformed CSV string to a row with all `null`s in the PERMISSIVE mode. Since Spark 3.0, returned row can contain non-`null` fields if some of CSV column values were parsed and converted to desired types successfully.
 
+  - In Spark version 2.4 and earlier, JSON datasource and JSON functions like `from_json` convert a bad JSON record to a row with all `null`s in the PERMISSIVE mode when specified schema is `StructType`. Since Spark 3.0, returned row can contain non-`null` fields if some of JSON column values were parsed and converted to desired types successfully.
+
```

Comment:
hmm, if returned row contains non null fields, how do we know if the row is read from a bad JSON record or a correct JSON record?

Reply:
And this behavior is also defined at some places like DataFrameReader.

Reply:
If the row was produced from a bad JSON record, the bad record is placed to the corrupt column otherwise the corrupt column contains null.

Reply:
In the `PERMISSIVE` mode, no way but at the moment (without the PR) you cannot distinguish a row produced from a bad record from a row produced from JSON object with all `null` fields too.

A row itself with all null cannot be an indicator of bad record. Need an additional flag. `null` or non-`null` in the corrupt column plays such role.

Reply:
For such behavior change, shall we add a config to roll back to previous behavior?

Reply:
And you should also update other places where defines previous behavior, like DataFrameReader.